### DISCUSSION TITLE
[Feature Request] NTLM Proxy Authentication with credentials.

### DISCUSSION THREAD

**Discussion by chahilp** [AzureAD/microsoft-authentication-library-for-python#557] [Q&A] [answered]
There is a missing option to authenticate the proxy with credentials.

What I am trying to do?
1. I am trying to generate access token for the Graph API calls with oauth2.0.
2. I am using AAD client credentials flow under corporate proxy.

To be able to access external website such as https://microsoft.com or the api endpoints to generate access token requires proxy authentication with credentials which is being fulfilled by requests module in Python by **auth=("proxyusername", "myproxypassword")**, such similar option is being expected to authenticate proxy to reach microsoft cloud authentication endpoints to generate access toke.

**Comment by rayluo** [comment, marked as answer]
Thanks for your patience.

MSAL Python provided [some parameters and relay them to the underlying `requests`](https://github.com/AzureAD/microsoft-authentication-library-for-python/blob/1.20.0/msal/application.py#L272-L286), but that approach is not scalable. So, nowadays, MSAL Python allows app developer to provide a [customized `http_client`](https://msal-python.readthedocs.io/en/latest/#msal.PublicClientApplication.params.http_client) that can have all the behavior you choose to have.

If you just want to customize the proxy behavior, you may just rely on [`requests`'s environmental variable behavior](https://requests.readthedocs.io/en/latest/user/advanced/#proxies).
